### PR TITLE
DM-9833: Create new sourceSelector BaseClass

### PR DESCRIPTION
**Pull request by morriscb**
This pull request is for review only. It will not be merged into the stack. The new source selector base class will be merged once all of the source selectors currently implemented in the stack are updated to the new API.

### CODE REVIEW

File: tests/testFlaggedSourceSelector.py [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+
+        self.src = afwTable.SourceCatalog(schema)
+        self.sourceSelector = \
+            sourceSelector.sourceSelectorRegistry['flagged'](schema=schema)
+
+    def tearDown(self):
+        del self.src
+        del self.sourceSelector
+
+    def testSelectSources_good(self):
```
Comment (djreiss) [inline tests/testFlaggedSourceSelector.py]: Docstring here would be useful.

File: tests/testFlaggedSourceSelector.py [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+        for i in range(5):
+            add_good_source(self.src, i)
+        result = self.sourceSelector.run(self.src)
+        # TODO: assertEqual doesn't work correctly on source catalogs.
+        # self.assertEqual(result.sourceCat, self.src)
+        for x in self.src['id']:
+            self.assertIn(x, result.source_cat['id'])
+
+    def testSelectSources_selected_field(self):
```
Comment (djreiss) [inline tests/testFlaggedSourceSelector.py]: Docstring here too.

File: tests/testFlaggedSourceSelector.py [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+        self.src[0].set("calib_psfUsed", False)
+        result = self.sourceSelector.run(
+            self.src, source_selected_field=self.selected_key)
+        self.assertFalse(self.src[0].get("is_selected"))
+        self.assertTrue(self.src[1].get("is_selected"))
+        self.assertTrue(self.src[2].get("is_selected"))
+        self.assertTrue(self.src[3].get("is_selected"))
+        self.assertTrue(self.src[4].get("is_selected"))
+
+    def testSelectSources_bad(self):
```
Comment (djreiss) [inline tests/testFlaggedSourceSelector.py]: Docstring please.

File: tests/testFlaggedSourceSelector.py [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+            self.assertIn(x, result.source_cat['id'])
+
+    def testSelectSources_selected_field(self):
+        for i in range(5):
+            add_good_source(self.src, i)
+        self.src[0].set("calib_psfUsed", False)
+        result = self.sourceSelector.run(
+            self.src, source_selected_field=self.selected_key)
+        self.assertFalse(self.src[0].get("is_selected"))
+        self.assertTrue(self.src[1].get("is_selected"))
```
Comment (djreiss) [inline tests/testFlaggedSourceSelector.py]: Put in a loop?

File: python/lsst/meas/algorithms/flaggedSourceSelector.py [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+    def select_sources(self, source_cat, masked_image=None, matches=None):
+        """Return a bool array representing which sources to select from
+        source_cat.
+
+        The input catalog must be contiguous in memory.
+
+        Parameters
+        ----------
+        source_cat : lsst.afw.table.SourceCatalog
+            catalog of sources that may be sources
```
Comment (morriscb) [inline python/lsst/meas/algorithms/flaggedSourceSelector.py]: Changed to "Catalog of sources to select from."

File: python/lsst/meas/algorithms/flaggedSourceSelector.py [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+        masked_image : {None} lsst.afw.image
+            An image containing the sources tests or for plotting.
```
Comment (morriscb) [inline python/lsst/meas/algorithms/flaggedSourceSelector.py]: Changed to "An image containing the sources for use in selection tests or for plotting."

File: python/lsst/meas/algorithms/flaggedSourceSelector.py [hunk 7]
```diff
@@ -0,0 +1,96 @@
...
+        ------
+        lsst.pipe.base.Struct
+            The struct contains the following data:
+
+            selected : bool array
+                Boolean array of sources that were selected, same length as
+                source_cat.
+        """
+        return pipeBase.Struct(
+            selected=source_cat.get(self.key),)
```
Comment (morriscb) [inline python/lsst/meas/algorithms/flaggedSourceSelector.py]: See RFC-335, source selectors can now assume a contiguous catalog.

File: python/lsst/meas/algorithms/sourceSelector.py [hunk 8]
```diff
@@ -1,98 +1,138 @@
...
 import lsst.pex.config as pexConfig
 import lsst.pipe.base as pipeBase
 from future.utils import with_metaclass
 
+__all__ = ["BaseSourceSelectorConfig", "BaseSourceSelectorTask",
+           "sourceSelectorRegistry"]
+
 
 class BaseSourceSelectorConfig(pexConfig.Config):
-    badFlags = pexConfig.ListField(
```
Comment (morriscb) [inline python/lsst/meas/algorithms/sourceSelector.py]: Sorry it wasn't clearer. This is a implementation ticket for RFC-198 so we are changing the API. This ticket is mainly to create and test the new base class. The two selectors you mention will be updated in DM-9835 and DM-10797 respectively.

File: python/lsst/meas/algorithms/sourceSelector.py [hunk 9]
```diff
@@ -1,98 +1,138 @@
...
-        ],
-    )
+    pass
 
 
 class BaseSourceSelectorTask(with_metaclass(abc.ABCMeta, pipeBase.Task)):
-    """!Base class for source selectors
+    """Base class for source selectors
+
+    Write paragraph about what a source selector is/does.
```
Comment (morriscb) [inline python/lsst/meas/algorithms/sourceSelector.py]: Thanks, for catching this. I have added one.

File: python/lsst/meas/algorithms/sourceSelector.py [hunk 10]
```diff
@@ -1,98 +1,138 @@
...
-        """!Return a catalog of sources: a subset of sourceCat.
-
-        @param[in] sourceCat  catalog of sources that may be sources (an lsst.afw.table.SourceCatalog)
-
-        @return a pipeBase.Struct containing:
-        - sourceCat  a catalog of sources
+    def select_sources(self, source_cat, masked_image=None, matches=None):
+        """Return a catalog of sources selected by specified criteria.
+
+        The input catalog must be contiguous in memory.
```
Comment (morriscb) [inline python/lsst/meas/algorithms/sourceSelector.py]: Yes, this is part of RFC-335.

File: tests/testFlaggedSourceSelector.py [hunk 11]
```diff
@@ -0,0 +1,123 @@
...
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the LSST License Statement and
+# the GNU General Public License along with this program.  If not,
+# see <https://www.lsstcorp.org/LegalNotices/>.
+#
+from __future__ import division, absolute_import, print_function
```
Comment (morriscb) [inline tests/testFlaggedSourceSelector.py]: Isn't this part of the standard boiler plate for backwards compatibility with python2.7?

File: python/lsst/meas/algorithms/astrometrySourceSelector.py [hunk 12]
```diff
@@ -43,6 +43,18 @@ class AstrometrySourceSelectorConfig(BaseSourceSelectorConfig):
         "(in the flux specified by sourceFluxType); <= 0 for no limit",
         default=10,
     )
+    bad_flags = pexConfig.ListField(
```
Comment (morriscb) [inline python/lsst/meas/algorithms/astrometrySourceSelector.py]: Huh, I didn't think I touched this file. This was an unintentional change. I'll revert it back to the version that is on master.

**Comment by morriscb** [comment]
Committed responses to the first round of reviewer comments.

File: tests/testFlaggedSourceSelector.py [hunk 13]
```diff
@@ -77,18 +79,27 @@ def testSelectSources_good(self):
             self.assertIn(x, result.source_cat['id'])
 
     def testSelectSources_selected_field(self):
+        """Test the behavior of source_selected_field in run.
+
+        This test asserts that the field will specified in source selected
+        field is properly set by the source selector. We test this both for
```
Comment (djreiss) [inline tests/testFlaggedSourceSelector.py]: "source selected field" -> `source_selected_field`  ?

File: tests/testFlaggedSourceSelector.py [hunk 14]
```diff
@@ -77,18 +79,27 @@ def testSelectSources_good(self):
...
             add_good_source(self.src, i)
         self.src[0].set("calib_psfUsed", False)
         result = self.sourceSelector.run(
             self.src, source_selected_field=self.selected_key)
-        self.assertFalse(self.src[0].get("is_selected"))
-        self.assertTrue(self.src[1].get("is_selected"))
-        self.assertTrue(self.src[2].get("is_selected"))
-        self.assertTrue(self.src[3].get("is_selected"))
-        self.assertTrue(self.src[4].get("is_selected"))
+        for src_idx, soruce in enumerate(self.src):
```
Comment (djreiss) [inline tests/testFlaggedSourceSelector.py]: soruce -> source. Actually you don't use 'soruce' and therefore don't need enumerate.

File: tests/testFlaggedSourceSelector.py [hunk 15]
```diff
@@ -77,18 +79,27 @@ def testSelectSources_good(self):
...
+            if src_idx == 0:
+                self.assertFalse(self.src[src_idx].get("is_selected"))
+            else:
+                self.assertTrue(self.src[1].get("is_selected"))
```
Comment (djreiss) [inline tests/testFlaggedSourceSelector.py]: [1] -> [src_idx], I think?

**Comment by morriscb** [comment]
Fixed second round of requests. Also added new ticket branch tickets/DM-9832 that will collect work for this epic.

**Comment by djreiss** [comment]
Looks good to me. Feel free to Jenkins it to make sure it all works prior to merging (once you get around to merging).

**Comment by morriscb** [comment]
Thanks. I'll close this pull request and merge the code into tickets/DM-9832 as to merge this code to master requires more work and updates to the other source selectors throughout the stack.

**Pull request closed without merging**